### ISSUE TITLE
A way to wait for the data from the useQuery hook

### ISSUE THREAD

**Issue by mwanago** [reduxjs/redux-toolkit#3104]
Hi. I've been using RTK for quite some time now, and it's great. Thank you for all of your hard work. I have a question.

I'm running in the following case very often:
```javascript
function usePosts() {
  const { data } = useGetPostsQuery();
  const [createNewPost] = useCreatePostMutation();
  
  const saveNewPost(newPostTitle) {
    // I don't know if the data is loaded yet
    const matchingPost = data.find((post) => {
      return post.title === newPostTitle
    });
    
    if (matchingPost) {
      console.log('A post with this title already exists');
      return;
    }
    
    createNewPost(newPostTitle);
  }
  
  return {
    data,
    saveNewPost
  }
}
```

In the `saveNewPost1` function above, I don't know if the `data` is loaded yet. I know that I could use a lazy query, but I still want my `usePosts` hook to return the data.

The perfect solution would be something like the following:
```javascript
function usePosts() {
  const { data, dataPromise } = useGetPostsQuery();
  const [createNewPost] = useCreatePostMutation();
  
  const saveNewPost(newPostTitle) {
    const posts = await dataPromise;
  
    const matchingPost = posts.find((post) => {
      return post.title === newPostTitle
    });
    
    if (matchingPost) {
      console.log('A post with this title already exists');
      return;
    }
    
    createNewPost(newPostTitle);
  }
  
  return {
    data,
    saveNewPost
  }
}
```

Is that possible, or do you maybe plan on adding such a feature?
Thanks and take care.

**Comment by mwanagosos** [comment]
@markerikson Would it make sense to expose the latest promise in such a way?

**Comment by phryneas** [comment]
> In the saveNewPost1 function above, I don't know if the data is loaded yet.

You... do know if it is loaded... take a look at `isLoading`. The component will rerender every time that changes and you will have a new instance of `saveNewPost` every time.

> I know that I could use a lazy query, but I still want my usePosts hook to return the data.

You have to explain this a bit... why wouldn't your hook return data?

> Would it make sense to expose the latest promise in such a way?

No, generally we do not plan on doing that - it leads to patterns that we generally do not want to encourage.



Honestly, I'm just very confused by the issue you seem to have here. You might need to explain a bit more.

**Comment by mwanagosos** [comment]
Hi @phryneas. Let me rephrase:
I would like to have a way to wait for the data from the `useQuery` hook using a promise.
A good example is when a user clicks a button while the data is not yet accessible.

My requirements are:
- don't load the posts data if it has been loaded already
- if the user clicks on a button before the data has been loaded, wait for the request to finish

I could use a lazy query instead to have more control over the request, but that would mean I would have to reject a lot of functionalities the `useQuery` hook provides out of the box.
```
function usePosts() {
  const [getPosts, data] = useLazyGetPostsQuery();
  const [createNewPost] = useCreatePostMutation();
  const getPostsPromise = useRef();
  
  useEffect(() => {
    getPostsPromise.current = getPosts().unwrap();
  }, [getPosts]);
  
  async function saveNewPost(newPostTitle) {
    const posts = await getPostsPromise.current;
    
    const matchingPost = posts.find((post) => {
      return post.title === newPostTitle
    });
    
    if (matchingPost) {
      console.log('A post with this title already exists');
      return;
    }
    
    createNewPost(newPostTitle);
  }
  
  return {
    data,
    saveNewPost
  }
}
```
The above approach will always load a list of posts, even if they were previously loaded. 
Is there a better solution to this issue?

**Comment by mwanagosos** [comment]
@phryneas I tried using `getRunningQueryThunk`, but it does not seem to solve this problem.

**Comment by mwanagosos** [comment]
Looks like the `preferCacheValue` argument to the trigger function of the lazy query solves this problem for me completely.
```
function usePosts() {
  const { data } = useGetPostsQuery();
  const [getPosts] = useLazyGetPostsQuery();
  const [createNewPost] = useCreatePostMutation();
  
  async function saveNewPost(newPostTitle) {
    const posts = await getPosts(undefined, true).unwrap();
  
    const matchingPost = posts.find((post) => {
      return post.title === newPostTitle
    });
    
    if (matchingPost) {
      console.log('A post with this title already exists');
      return;
    }
    
    createNewPost(newPostTitle);
  }
  
  return {
    data,
    saveNewPost
  }
}
```
Correct me if I'm wrong, but it looks like `useLazyGetPostsQuery` and `useGetPostsQuery` shares the same cache that I'm able to access thanks to the second argument to the trigger function. Thanks to that, I'm able to achieve the behavior I need :tada: 
Thanks a lot for participating and have an awesome week.

**Comment by markerikson** [comment]
@mwanagosos : correct. There's only _one_ cache for an API, and the various things like `useQuery`, `useLazyQuery`, `dispatch(someEndpoint.initiate())`, etc, all interact with that one same cache.

**Comment by mwanagosos** [comment]
@markerikson Thanks for the confirmation, I appreciate it. I'll take advantage of that.
